PPL geoip function

kenrickyap opened the pull request:
Description
PPL geoip function

Issues Resolved
https://github.com/opensearch-project/opensearch-spark/issues/672

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

This PR is a continuation of https://github.com/opensearch-project/opensearch-spark/pull/781 due to lacking permissions to push to forked branch in said PR

kenrickyap commented:
Hi @YANG-DB, I heard from Anas that you had a method of implementing ip2geo functionality for Spark. I wanted to check with you that our current approach aligns with your method.

Current Plan:
Leveraging [SerializableUdf](https://github.com/opensearch-project/opensearch-spark/pull/706/files#diff-4fc2e6dc7264b8eac89050f1f19c51381f61fa36b302d78bcc0bc260498845fe) create a UDF that does the follow:

1. Check if in-memory cache object for datasource exists.
2. If cache object does not exists create new in-memory cache object from csv retrieved from datasource manifest.
3. Search cached object for GeoIP data.
4. Return GeoIP data.

All of this would have to be implemented within the Spark library, as currently I am not aware of how to access any geospatial artifacts. If you know of a better way to implement ip2geo please let me know! Thanks!

kenrickyap commented on ppl-spark-integration/src/main/java/org/opensearch/sql/expression/function/SerializableUdf.java:
```diff
@@ -51,7 +55,28 @@ public Boolean apply(String ipAddress, String cidrBlock) {
+        public Object apply(String datasource, String ipAddress, List<String> properties) {
+            Object results = "geoip data";
+
+            //TODO:
```
In-memory ip2geo functionality will be implemented here

YANG-DB replied:
@kenrickyap can you please explain in more details the flow of the events from the beginning of the query?

YANG-DB commented:
@kenrickyap please update the DCO (contributor sign-off)

YANG-DB commented on ppl-spark-integration/src/main/antlr4/OpenSearchPPLParser.g4:
```diff
@@ -449,6 +449,7 @@ valueExpression
 
 primaryExpression
```
@kenrickyap please review [this](https://github.com/opensearch-project/opensearch-spark/blob/p1-antlr-grammar/ppl-spark-integration/src/main/antlr4/OpenSearchPPLParser.g4) antlr p1 suggestion and see how the geoIp command is used as part of the `eval` reusing evalFunctionCall...
IMO IP function sounds not one of the primary expressions.

YANG-DB commented on ppl-spark-integration/src/main/java/org/opensearch/sql/ppl/CatalystExpressionVisitor.java:
```diff
@@ -33,31 +33,9 @@
-import org.opensearch.sql.ast.expression.When;
-import org.opensearch.sql.ast.expression.WindowFunction;
-import org.opensearch.sql.ast.expression.Xor;
+import org.opensearch.sql.ast.expression.*;
```
please keep the existing explicit import list ...

YANG-DB reviewed (changes requested):
@kenrickyap I'm missing a more detailed design of the actual functionality that should include:
 - flow diagram of loading the geoIp data
 - storage of that data
 - utilizing of different `datasource`
 - pro/cons for the different approaches 

please add the former discussions [made](https://github.com/opensearch-project/opensearch-spark/pull/781) into this issue for better traceability

kenrickyap commented:
**Flow diagram for ip2geo UDF**

![mermaid-diagram-2024-11-06-140042](https://github.com/user-attachments/assets/a7d48707-0348-4558-9b45-8a51a353d1d8)

**Design Details**

- Ip2GeoCache will be a Gauva Cache that has datasource string as key and CidrTree as value
- CidrTree will be Trie (will use apache.common PatriciaTrie as I see that apache.common is already included in project)
  - Am using trie instead of map as it is more well suited for longest prefix matching task, and this task is similar to ip to cidr matching
  - each CidrTreeNode will store:
    - nth bit value of cidr (n is the depth of tree)
    - geo_data if there is matching cidr row in 
    - child CidrTreeNodes
 - Will retrieve CsvParser from manifest using similar methodology as in geospatial [ip2geo](https://github.com/opensearch-project/geospatial/blob/main/src/main/java/org/opensearch/geospatial/ip2geo/dao/GeoIpDataDao.java#L188)
 
 **Pros**
 
 - ip2geo functionality is achieved.
 - implementation is simple and does not depend on any additional libraries that do not already exist in the project.
 
 **Cons**
 
 - calculations are done in-memory as a UDF, this means that multiple instances of Ip2GeoCache will be created in distributed Spark systems and they will not sync.
 - not leveraging job-scheduler to run ip2geo task and not leveraging OpenSearch to store ip2geo data as geospatial does.

YANG-DB commented:
> **Flow diagram for ip2geo UDF**
> 
> ![mermaid-diagram-2024-11-06-140042](https://private-user-images.githubusercontent.com/121634635/383740205-a7d48707-0348-4558-9b45-8a51a353d1d8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA5MzcxNTksIm5iZiI6MTczMDkzNjg1OSwicGF0aCI6Ii8xMjE2MzQ2MzUvMzgzNzQwMjA1LWE3ZDQ4NzA3LTAzNDgtNDU1OC05YjQ1LThhNTFhMzUzZDFkOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEwNlQyMzQ3MzlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03ZGE4ZTM3NzE5NDU1Y2NlOWNhNjVhNDFjOWFiZDBiZWJjZWIzOWRkYzNmMDllMmM5MmFhY2Y0ZDJmNzZkNjI1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.xQmnmdwvtN2M2Dt0INzk-OtazcQRuErwn7ewiaKvRGg)
> 
> **Design Details**
> 
> * Ip2GeoCache will be a Gauva Cache that has datasource string as key and CidrTree as value
> * CidrTree will be Trie (will use apache.common PatriciaTrie as I see that apache.common is already included in project)
>   
>   * Am using trie instead of map as it is more well suited for longest prefix matching task, and this task is similar to ip to cidr matching
>   * each CidrTreeNode will store:
>     
>     * nth bit value of cidr (n is the depth of tree)
>     * geo_data if there is matching cidr row in datasource csv
>     * child CidrTreeNodes
> * Will retrieve CsvParser from manifest using similar methodology as in geospatial [ip2geo](https://github.com/opensearch-project/geospatial/blob/main/src/main/java/org/opensearch/geospatial/ip2geo/dao/GeoIpDataDao.java#L188)
> 
> **Pros**
> 
> * ip2geo functionality is achieved.
> * implementation is simple and does not depend on any additional libraries that do not already exist in the project.
> 
> **Cons**
> 
> * calculations are done in-memory as a UDF, this means that multiple instances of Ip2GeoCache will be created in distributed Spark systems and they will not sync.
> * not leveraging job-scheduler to run ip2geo task and not leveraging OpenSearch to store ip2geo data as geospatial does.

@kenrickyap a few questions here:
- how / is the Ip2GeoCache memory shared between sessions ? is it a signleton ? 
- plz also add the use case where the geo_data_source will be a table / index (in OpenSearch) / service (API) - lets create a general purpose facade for this to hide different datasource drivers
- I would like to see a more detailed description of the `CidrTreeNode` including an example and simple high level explanations (the geo tree diagram ?)
- explain why is it worth while to add this trie-tree instead of using a hash-map / search ?
- can u also add a pseudo-code here (in the issue) to clarity the composition ?

kenrickyap commented:
> @kenrickyap a few questions here:
> 
> * how / is the Ip2GeoCache memory shared between sessions ? is it a signleton ?

Yes we will create Ip2GeoCache to be a singleton. From my understanding this should allow the cache to be accessible between sessions 

> * plz also add the use case where the geo_data_source will be a table / index (in OpenSearch) / service (API) - lets create a general purpose facade for this to hide different datasource drivers

Will update the flow diagram to reflect facade for different datasource. 

However, would it be possible to provide a service (API) example usage? I am not to sure what such a datasource would be expected to return. 

Also would there be a fixed schema for an index in opensource?

> * I would like to see a more detailed description of the `CidrTreeNode` including an example and simple high level explanations (the geo tree diagram ?)

In hindsight I will just use a hashmap to store cidr geo_data where the cidr bitstring will be the key and geo_data will be a value. To preform the ip cidr matching will implement lookup function to convert ip to bitstring and reduce bit length till key is found in map.

Initially I wanted to leverage the `prefixMap` function as I thought this would find the best fitting cidr mask for a given ip in O(1), which would mean I wouldn't have to implement my own lookup function. However as I was trying implement this noticed that `prefixMap` takes a prefix and finds all keys that have the prefix, which is the opposite with what I want.

> * explain why is it worth while to add this trie-tree instead of using a hash-map / search ?

As mentioned above will use a hash-map instead of trie.

> * can u also add a pseudo-code here (in the issue) to clarity the composition ?

Will have pseudo-code and code added by EOD

normanj-bitquill commented on docs/ppl-lang/planning/ppl-geoip.md:
```diff
@@ -0,0 +1,39 @@
+
+geoip function to add information about the geographical location of an IPv4 or IPv6 address
+
+1. **Proposed syntax**
```
Perhaps also list out the values that can be used for properties.

YANG-DB replied:
we need a larger set of examples here plz

kenrickyap replied:
added additional examples in ppl_ip.md doc

normanj-bitquill commented on ppl-spark-integration/src/main/antlr4/OpenSearchPPLParser.g4:
```diff
@@ -894,6 +895,10 @@ coalesceFunctionName
    ;
 
+geoIpFunctionCall
+   : GEOIP LT_PRTHS (datasource = functionArg COMMA)? ipAddress = functionArg (COMMA properties = stringLiteral)? RT_PRTHS
```
You could do further parsing here of `properties`. Something like:
```
... (COMMA properties =  geoipPropertiesList)? ...

geoipPropertiesList
  : geoipProperty (COMMA geoipProperty)*

geoipProperty
  : city
  | lat
  | lon
```

YANG-DB replied:
I like this idea !!

normanj-bitquill commented on ppl-spark-integration/src/main/java/org/opensearch/sql/ast/AbstractNodeVisitor.java:
```diff
@@ -5,41 +5,11 @@
-import org.opensearch.sql.ast.expression.Field;
-import org.opensearch.sql.ast.expression.FieldList;
-import org.opensearch.sql.ast.expression.LambdaFunction;
+import org.opensearch.sql.ast.expression.*;
```
Avoid using `*` imports.

normanj-bitquill commented on ppl-spark-integration/src/test/scala/org/opensearch/flint/spark/ppl/PPLLogicalPlanEvalTranslatorTestSuite.scala:
```diff
@@ -177,6 +177,10 @@ class PPLLogicalPlanEvalTranslatorTestSuite
   }
 
+  test("test eval expression - geoip function") {
+
```
Should add a TODO comment here, since it appears you intend to fill this out later.

YANG-DB commented on ppl-spark-integration/src/main/java/org/opensearch/sql/common/geospatial/CidrGeoMap.java:
```diff
@@ -0,0 +1,41 @@
+public class CidrGeoMap extends HashMap<String, GeoIpData> {
+
+    private String toBinaryString(String ip) {
+        StringBuilder binary = new StringBuilder();
```
@kenrickyap please use the [ip library](https://github.com/opensearch-project/opensearch-spark/blob/dd9c0cfbab0ca327ba655c8aa9f9ee44404d4fce/build.sbt#L188C6-L188C55) for such text to utilis

YANG-DB reviewed (changes requested):
@kenrickyap can u plz update the PR status and progress ?
thanks

YANG-DB commented:
@kenrickyap plz fix the DCO error ...

YANG-DB commented on ppl-spark-integration/src/main/java/org/opensearch/sql/common/geospatial/CidrGeoMap.java:
```diff
@@ -0,0 +1,63 @@
+        return res;
+    }
+
+    private void put(String cidr, GeoIpData data) throws UnknownHostException {
```
@kenrickyap can we use an IP util library for these string to ip conversions ?

YANG-DB commented on ppl-spark-integration/src/main/java/org/opensearch/sql/common/geospatial/DatasourceDao.java:
```diff
@@ -0,0 +1,37 @@
+    Stream<Pair<BitSet, GeoIpData>> getGeoIps();
+
+    static BitSet createCidrBitSet(String cidr) {
+        String[] parts = cidr.split("/");
```
same here

YANG-DB commented on ppl-spark-integration/src/main/java/org/opensearch/sql/common/geospatial/TestDatasourceDao.java:
```diff
@@ -0,0 +1,35 @@
+        // Mock GeoIpData entries
+        GeoIpData geoIp1 = new GeoIpData("US", "United States", "North America", "US-CA", "California", "Los Angeles", "America/Los_Angeles", "34.0522", "-118.2437");
+        GeoIpData geoIp2 = new GeoIpData("CA", "Canada", "North America", "CA-ON", "Ontario", "Toronto", "America/Toronto", "43.65107", "-79.347015");
+        BitSet bitSet1 = DatasourceDao.createCidrBitSet("192.168.0.0/24"); // Example CIDR mask
```
more test cases needed including ipv6 and sample errors in the ip

YANG-DB commented:
@LantaoJin I would appreciate u'r review here as well...

LantaoJin commented on ppl-spark-integration/src/main/java/org/opensearch/sql/common/geospatial/TestDatasourceDao.java:
```diff
@@ -0,0 +1,35 @@
+import java.util.BitSet;
+import java.util.stream.Stream;
+
+public class TestDatasourceDao implements DatasourceDao {
```
why a test class here?

jduo replied:
This will be moved... DataSourceDaoFactory depends on this currently and will be updated to load implementations with reflection in an update on this PR.

LantaoJin commented on ppl-spark-integration/src/main/java/org/opensearch/sql/common/geospatial/DatasourceManifest.java:
```diff
@@ -0,0 +1,94 @@
+   * @param sha256Hash SHA256 hash value of a database file
+   * @return SHA256 hash value of a database file
+   */
+  @JsonProperty("sha256_hash")
```
Seems this annotation is the only spot to use jackson?

jduo replied:
ObjectMapper is from Jackson

jduo replied:
If we're OK using ObjectMapper because it's provided through AWS dependencies, I can remove use of @JsonProperty to get rid of Jackson databind.

LantaoJin commented on ppl-spark-integration/src/main/java/org/opensearch/sql/common/geospatial/DatasourceDaoFactory.java:
```diff
@@ -0,0 +1,17 @@
+package org.opensearch.sql.common.geospatial;
```
missing header.

LantaoJin commented on build.sbt:
```diff
@@ -155,6 +155,10 @@ lazy val pplSparkIntegration = (project in file("ppl-spark-integration"))
       "com.github.seancfoley" % "ipaddress" % "5.5.1",
+      "org.apache.commons" % "commons-lang3" % "3.17.0",
+      "org.apache.commons" % "commons-csv" % "1.12.0",
+      "org.apache.commons" % "commons-csv" % "1.12.0",
```
is it a duplicated line?

jduo replied:
Yes looks to be a typo.

LantaoJin commented on ppl-spark-integration/src/main/java/org/opensearch/sql/common/geospatial/GeoIpCache.java:
```diff
@@ -0,0 +1,31 @@
+
+    private GeoIpCache() {
+        cache = CacheBuilder.newBuilder()
+                .expireAfterWrite(3, TimeUnit.DAYS)
```
The TTL should be configurable

jduo replied:
This implies the same data source can have different IPs depending on user context... is this something possible @YANG-DB ?

YANG-DB replied:
@jduo yes it is possible - lets add a use case + document this

jduo replied:
This really can't be a singleton in that case. We need to store this with the session/user context. How might we access this @YANG-DB ? Or we make it a multi-level cache or use a composite key for the cache.

LantaoJin commented on ppl-spark-integration/src/main/java/org/opensearch/sql/common/geospatial/TestDatasourceDao.java:
```diff
@@ -0,0 +1,35 @@
+package org.opensearch.sql.common.geospatial;
```
missing header

jduo commented on build.sbt:
```diff
@@ -155,6 +155,10 @@ lazy val pplSparkIntegration = (project in file("ppl-spark-integration"))
+      "org.apache.commons" % "commons-lang3" % "3.17.0",
+      "org.apache.commons" % "commons-csv" % "1.12.0",
+      "org.apache.commons" % "commons-csv" % "1.12.0",
+      "com.fasterxml.jackson.core" % "jackson-annotations" % "2.14.2",
```
Hi @LantaoJin ,
What's the best way for this project to parse JSON?

andy-k-improving replied:
We can probably use the Jackson dep, but with the version that defined on line: 12, with runtime only scope.
```
lazy val jacksonVersion = "2.15.2"
```

jduo commented on ppl-spark-integration/src/main/java/org/opensearch/sql/common/geospatial/GeoIpCache.java:
```diff
@@ -0,0 +1,31 @@
+
+    public Cache<String, CidrGeoMap> cache;
+
+    private static GeoIpCache cacheInstance = null;
```
This should be final.
The singleton pattern can be simplified in Java to have no getInstance() method:
```suggestion
    private static GeoIpCache cacheInstance = null;
    private static final GeoIpCache INSTANCE = new GeoIpCache();
```

jduo commented on ppl-spark-integration/src/main/java/org/opensearch/sql/common/geospatial/ManifestDao.java:
```diff
@@ -0,0 +1,114 @@
+
+  @Override
+  public Stream<Pair<BitSet, GeoIpData>> getGeoIps() {
+    manifestCsv = getDatabaseReader(manifest);
```
This can be moved to the constructor and made final

qianheng-aws commented:
> * Ip2GeoCache will be a Gauva Cache that has datasource string as key and CidrTree as value
> *

questions about the design:

1. The UDF will only be executed on Spark Executor. So the cache won't take effect actually, it will be destroyed once the task has finished. Do we have mechanism to avoid this?
2. The CSV data is nearly 500MB for the datasource of `https://geoip.maps.opensearch.org/v1/geolite2-city/manifest.json`, I guess it will get bigger if populating a trie tree on it. It think it will be a big memory pressure for Spark Executor.
3. Have we ever considered retrieving the geo data by calling the related API of service like OpenSearch [geospatial](https://github.com/opensearch-project/geospatial)? Then we don't have to load the above CSV in executor memory. But I'm not quite sure whether it's feasible to do that on Serverless Spark.

kenrickyap commented:
> > * Ip2GeoCache will be a Gauva Cache that has datasource string as key and CidrTree as value
> 
> @kenrickyap questions about the design:
> 
> 1. The UDF will only be executed on Spark Executor. So the cache won't take effect actually, it will be destroyed once the task has finished. Do we have mechanism to avoid this?

The JVM will persist on the worker node, so udf tasks that are carried out in the same worker node will be able to leverage the cache. In any case from above comments we will be shifting away from using a singleton and implement some other method of persisting queried results.

> 2. The CSV data is nearly 500MB for the datasource of `https://geoip.maps.opensearch.org/v1/geolite2-city/manifest.json`, I guess it will get bigger if populating a trie tree on it. I think it will be a big memory pressure for Spark Executor.

yes this will be the case, this would be an initial method of performing geo2ip, there is plan to add less expensive methods of performing operation in future.

> 3. Have we ever considered retrieving the geo data by calling the related API of service like OpenSearch [geospatial](https://github.com/opensearch-project/geospatial)? Then we don't have to load the above CSV in executor memory. But I'm not quite sure whether it's feasible to do that on Serverless Spark.

yes the initial plan was to use the geospatial plugin, but geospatial does not publish any jars, and I could not find a way to leverage it's functionality for Spark. A client jar for the plugin is currently being created, a plan is discussed here: **https://github.com/opensearch-project/sql/issues/3038**. The plan was to have a follow up task where we leverage the use of this client, when it is implemented, however every as there was an urgent ask for this functionality, we opted to implement an in-engine method in parallel.

There has also been ask to be able to retrieve data from OpenSearch Index and general APIs, these methods will also be added in future issue

LantaoJin commented:
>> 1. The UDF will only be executed on Spark Executor. So the cache won't take effect actually, it will be destroyed once the task has finished. Do we have mechanism to avoid this?

> The JVM will persist on the worker node, so udf tasks that are carried out in the same worker node will be able to leverage the cache.

Not exactly. It depends on how the user deploys Spark. In most cases, when the dynamic resource allocation enabled, when an executor is idle for a while, e.g. no active task is running on this executor in 60 seconds, the executor process (JVM) will be released. The cache won't take effect any more.

>> 2. The CSV data is nearly 500MB for the datasource of https://geoip.maps.opensearch.org/v1/geolite2-city/manifest.json, I guess it will get bigger if populating a trie tree on it. I think it will be a big memory pressure for Spark Executor.

> yes this will be the case, this would be an initial method of performing geo2ip, there is plan to add less expensive methods of performing operation in future.

What's the plan? I can't agree this initial method where each executor hosts huge memory and the cache doesn't take effect. At least the memory/cache should be built in a lazy mode, for users who doesn't use this geoip function.

LantaoJin commented:
My thought is that from the perspective of open-source project, we only need to provide an interface for using IP services and a configuration-based data source connector, but do not actually provide specific IP lookup services. Users need to configure or deploy IP data source services that they can use. For example, users need to complete the configuration and authorization of web services, such as https://dev.maxmind.com/geoip/geolocate-an-ip/web-services/
For AOS, an internal service for ip lookup should be deployed.

YANG-DB commented:
> My thought is that from the perspective of open-source project, we only need to provide an interface for using IP services and a configuration-based data source connector, but do not actually provide specific IP lookup services. Users need to configure or deploy IP data source services that they can use. For example, users need to complete the configuration and authorization of web services, such as https://dev.maxmind.com/geoip/geolocate-an-ip/web-services/
> For Amazon OpenSearch Service (AOS), an internal service for ip lookup should be deployed. 
> 
> Another thought is that we create two tables (IPv4 and IPv6) in Glue as builtin tables. Each IP lookup can be converted to a Join operator with one of the IP table.

I agree with @LantaoJin about the interface approach - but we can definitely support a default in-memory implementation for this to work out of the box

kenrickyap commented:
## GeoIP Implementation Options and Path Forward

**Current Status**
The in-engine implementation of GeoIP functionality has core features working, with remaining work focused on tests, parameter filtering, and caching improvements.

### Implementation Options

**Option 1: In-Engine Implementation**

**Approach**: Perform IP lookups within the engine using in-memory processing

**Pros**:

- Simpler deployment (no external dependencies)
- Lower latency for lookups
- Immediate availability after setup
- Consistent performance

**Cons**:

- Higher memory usage (~1.3GB JVM OnHeap per worker)
- Memory scales with worker nodes
- Limited flexibility in updating GeoIP database

**Option 2: External Service Integration**

**Approach**: Offload IP lookups to external services or database

**Pros**:

- Lower memory footprint on worker nodes
- More flexible updating of GeoIP database
- Scalable independent of worker nodes
- Potential for more accurate/updated data

**Cons**:

- Additional setup complexity for users
- External service dependencies
- Potential latency overhead
- Network bandwidth considerations

### Technical Implementation Details

**In-Engine Implementation**

- Uses ManifestDao for IP lookups
- Singleton cache for IP lookup table
- Direct integration with current geospatial module

**External Service Implementation**

- API datasource DAO for service integration
- Query-level result caching
- Configurable datasource options:
```
  geoip:
    datasource:
      type: “service|database”
      connection:
        # Service-specific configuration
        endpoint: “”
        credentials: “”
        # Or database-specific configuration
        table: “”
        index: “”
  ```
### Path Forward

**Proposed Approach**

We’re considering two potential paths:

1. **Phased Implementation:**
  - Implement in-engine solution first
  - Add external service support later
  - Allows for gradual transition
2. **Direct External Implementation:**
  - Focus directly on external service integration
  - Skip in-engine implementation
  - Faster path to scalable solution

### Next Steps

1. Complete current PR to release-ready state
2. Create RFC for API interface design
3. Present both options to community for feedback
4. Proceed based on community consensus

### Discussion Points

- Memory vs. latency tradeoffs
- Deployment complexity considerations
- Maintenance and updating strategies
- Migration path for existing users

### Seeking Feedback On

- Which path would be recommended moving forward?
- Design review on RFC for API interface design (Will provide RFC soon!)

andy-k-improving commented on docs/ppl-lang/planning/ppl-geoip.md:
```diff
@@ -0,0 +1,39 @@
+        - use the "abc" geoip datasource
+    - `... | eval geoinfo = geoip(ipAddress, "city,lat,lon")`
+        -  use the default geoip datasource, retrieve only city, lat and lon
+    - `... | eval geoinfo = geoip("abc", ipAddress, "city,lat,lon")`
```
Would be great to also list out all the supported attributes (ex: City, Lat.....etc) along with the respective data type.

kenrickyap replied:
added in ppl_ip.md doc

andy-k-improving commented on ppl-spark-integration/src/main/antlr4/OpenSearchPPLLexer.g4:
```diff
@@ -418,6 +418,19 @@ TYPEOF:                             'TYPEOF';
+//GEOLOCATION FUNCTIONS
+GEOIP:                              'GEOIP';
+
+//GEOLOCATION PROPERTIES
```
Will this list stay consistent across different implementation, or this is just for `ip2geo`?

YANG-DB replied:
@kenrickyap lets try to align according to this schema

kenrickyap replied:
current list is documented in docs and design files

andy-k-improving commented on ppl-spark-integration/src/main/java/org/opensearch/sql/common/geospatial/DatasourceManifest.java:
```diff
@@ -0,0 +1,98 @@
+   * Datasource manifest builder
+   */
+  public static class Builder {
+    private static final int MANIFEST_FILE_MAX_BYTES = 1024 * 8;
```
What the usage of this MAX_BYTES declaration?
We don't seem to have any reference for this var in the projec.

andy-k-improving commented on ppl-spark-integration/src/main/java/org/opensearch/sql/common/geospatial/DatasourceManifest.java:
```diff
@@ -0,0 +1,98 @@
+   * @param provider A database provider name
+   * @return A database provider name
+   */
+  public String provider;
```
Should we have this as enum instead?
So later on we can do a Java `switch` to route the enrichment request during runtime.

andy-k-improving commented on ppl-spark-integration/src/main/java/org/opensearch/sql/common/geospatial/GeoIpData.java:
```diff
@@ -0,0 +1,42 @@
+@Getter
+@ToString
+@Builder
+public class GeoIpData {
```
Will this data model specific to any data provider? 
Or this is generic for all future integration for new data source?

andy-k-improving commented on ppl-spark-integration/src/main/java/org/opensearch/sql/common/geospatial/ManifestDao.java:
```diff
@@ -0,0 +1,139 @@
+      }
+    });
+
+    int cidr_index = headerMap.get("cidr");
```
Will there be any null check on these columns? 
Also how's the handling will be like? (Ex: Treat it as corrupted file, and abort....etc)

jduo commented on ppl-spark-integration/src/main/java/org/opensearch/sql/common/geospatial/ManifestDao.java:
```diff
@@ -0,0 +1,139 @@
+   */
+  // TODO: Make this a configurable setting.
+  public static final String DATASOURCE_ENDPOINT =
+      "https://geoip.maps.opensearch.org/v1/geolite2-city/manifest.json";
```
The DATASOURCE_ENDPOINT variable isn't going to be used now. Instead the user will supply a manifest URL as part of the function call.

However downloading from user-supplied links is a security concern. This code is a simplified version of the manifest-reading code from the geospatial plugin:
https://github.com/opensearch-project/geospatial/blob/10a076a41cc6e3fe23061a198c1f55c1e1c402f8/src/main/java/org/opensearch/geospatial/ip2geo/dao/GeoIpDataDao.java#L174

Glancing at it, it seems there's some URL whitelisting and running the request in a privileged action. Would it be a good path going forward to replicate this behavior?

kenrickyap commented:
@YANG-DB question regarding key needed for user session caching, would this be information obtainable from the Spark instance?
Was looking through the environment variables and saw `hadoop.http.staticuser.user`, would this or some other spark env param be suitable? Another option is to have the user input some configuration on Spark start up.

YANG-DB commented:
> @YANG-DB question regarding key needed for user session caching, would this be information obtainable from the Spark instance? Was looking through the environment variables and saw `hadoop.http.staticuser.user`, would this or some other spark env param be suitable? Another option is to have the user input some configuration on Spark start up.

@kenrickyap we need to validate this with different spark providers such as the local out of the box spark  / EMR managed spark and such...

kenrickyap commented:
Based on discussion with @YANG-DB. Will shift implementation to only support Spark tables, where task of creating lookup table will be pushed to user. This shift will mean that there is no longer any reliance on UDF and performing ip2geo lookup in memory.

Specifically:
- geoip default behaviour is to perform join on predefined spark table. (Table name will be provided as config during spark startup.)
- pre-defined table schema is expected for geoip function to work. This will be documented.
- part of PR is to provide user with instructions to create spark geoip table.

We will provide ability to use other datasources for ip lookup in future tasks.

LantaoJin commented:
> Based on discussion with @YANG-DB. Will shift implementation to only support Spark tables (with intention to support other datasources in future), where task of creating lookup table will be pushed to user. This shift will mean that there is no longer any reliance on UDF and performing ip2geo lookup in memory.
> 
> Specifically:
> 
> * geoip default behaviour is to perform join on predefined spark table. (Table name will be provided as config during spark startup.)
> * pre-defined table schema is expected for geoip function to work. This will be documented.
> * part of PR is to provide user with instructions to create spark geoip table.
> 
> We will provide ability to use other datasources for ip lookup in future tasks.

Sounds good. I like this solution but it's still very challenging! The primary problem to solve in this solution is how to build a `Partitioner`: `HashPartitioner` doesn't work with CIDR notation. For example 1.0.1.0/24 includes all listings from 1.0.1.1 to 1.0.1.255, or we have to expand all IPv4 and IPv6 listings to a spark table.

kenrickyap commented:
@LantaoJin @YANG-DB This is the current details for the new implementation. Please let us know if there is any concerns or comments :)

Feature will include two parts:

- Scala script provided in docs that user will run to download data and create spark table.
- `geoip` function that will perform join on specified table.

### Details on Scala Script

#### User flow

- user will run script on spark-shell
- script will:
  1. download manifest from https://geoip.maps.opensearch.org/v1/geolite2-city/manifest.json and retrieve csv.
  2. parse csv and create spark dataframe.
  3. create new spark table `geoip_table` from dataframe.
    - table will have the following schema:
       cidr, country_iso_code ,country_name, continent_name, region_iso_code, region_name, city_name, time_zone, location
  4. add following columns to `geoip_table`: 
    - `ip_range_start` - type: `long` - first ip address in CIDR range (will ensure range is distinct does not overlap)
    - `ip_range_end` - type: `long` - last ip address in CIDR range (will ensure range is distinct does not overlap)
    - `ip_type` - type: `string` - IPv4 or IPv6
    
### Details of `geoip` implementation

`geoip` will have 3 inputs:
  - `datasource` - currently unused (will keep for future feature of letting users query API/Opensearch index datasources). 
  - `ip_address` - ip_address to get geo_info on
  - `parameters` - which columns will be selected from `geoip_table`.

#### ip2geo lookup details

To perform ip2geo look up we will perform join on `geoip_table` with the following conditions:

- geoip_table.ip_start_end <= ip_address <= geoip_table.ip_range_end
  - geo_table name will be provided at spark start up as config (e.g. --conf spark.geoip.tablename = `geoip_table`)
- ip_address.type = ip_type (will check IP type using inet library)

### Implementation clarification questions

- Regarding AWS EMR Spark, will the shell script be able to download manifest?
  - If not we can alter the script such that user needs to provide filepath to script, and offload csv retrieval to user
- Regarding @LantaoJin's suggestions to use `Partitioner`, I am not to sure how to integrate this given the current design. Currently my guess is that each node of the partition would be a different sub range, and we would query sub range directly to reduce load? Is this the right interpretation or am I missing something?

YANG-DB commented:
> @LantaoJin @YANG-DB This is the current details for the new implementation. Please let us know if there is any concerns or comments :)
> 
> Feature will include two parts:
> 
> * Scala script provided in docs that user will run to download data and create spark table.
> * `geoip` function that will perform join on specified table.
> 
> ### Details on Scala Script
> #### User flow
> * user will run script on spark-shell
> * script will:
>   
>   1. download manifest from https://geoip.maps.opensearch.org/v1/geolite2-city/manifest.json and retrieve csv.
>   2. parse csv and create spark dataframe.
>   3. create new spark table `geoip_table` from dataframe.
>   
>   
>   * table will have the following schema:
>     cidr, country_iso_code ,country_name, continent_name, region_iso_code, region_name, city_name, time_zone, location
>   
>   
>   4. add following columns to `geoip_table`:
>   
>   
>   * `ip_range_start` - type: `long` - first ip address in CIDR range (will ensure range is distinct and does not overlap)
>   * `ip_range_end` - type: `long` - last ip address in CIDR range (will ensure range is distinct and does not overlap)
>   * `ip_type` - type: `string` - IPv4 or IPv6
> 
> ### Details of `geoip` implementation
> `geoip` will have 3 inputs:
> 
> * `datasource` - currently unused (will keep for future feature of letting users query API/Opensearch index datasources).
> * `ip_address` - ip_address to get geo_info on
> * `parameters` - which columns will be selected from `geoip_table`.
> 
> #### ip2geo lookup details
> To perform ip2geo look up we will perform join on `geoip_table` with the following conditions:
> 
> * geoip_table.ip_start_end <= ip_address <= geoip_table.ip_range_end
>   
>   * geo_table name will be provided at spark start up as config (e.g. --conf spark.geoip.tablename = `geoip_table`)
> * ip_address.type = ip_type (will check IP type using inet library)
> 
> ### Implementation clarification questions
> * Regarding AWS EMR Spark, will the shell script be able to download manifest?
>   
>   * If not we can alter the script such that user needs to provide filepath to script, and offload csv retrieval to user
> * Regarding @LantaoJin's suggestions to use `Partitioner`, I am not to sure how to integrate this given the current design. Currently my guess is that each node of the partition would be a different sub range, and we would query sub range directly to reduce load on join? Is this the right interpretation or am I missing something?

Hi @kenrickyap 
Thanks for the detailed information 👍 here are my comments:
 - no need to download the [manifest-file](https://geoip.maps.opensearch.org/v1/geolite2-city/manifest.json) you should copy it locally to this repo (since its only an example and not a production use-case ).
 - A user can be expected to provide this file and use S3 file location for this purpose - we can instruct EMR users to manually upload the file into S3 and use the `create table ... using ... ` to load the data from there.
 - what is the expected size of this table ? (rows / memory) - plz test and verify
 - table structure: `ip_type` - type: `string` - IPv4 or IPv6 - why not use a bool ? (ipv4 columns)

YANG-DB commented:
@kenrickyap 
another request - please run `explain cost |  ip2geo ...`  to see how the query cost is estimated by the engine with resect to the amount of rows..,,

kenrickyap commented:
Thanks for the feedback :) have updated design to address comments. As for question responses:

> * what is the expected size of this table ? (rows / memory) - plz test and verify

Table has about 5 million rows, and is about 500 MB.

> * please run explain cost |  ip2geo ... to see how the query cost is estimated by the engine with relation to the amount of actual rows..

Will verify this when the command is implemented

LantaoJin commented:
> #### ip2geo lookup details
> To perform ip2geo look up we will perform join on `geoip_table` with the following conditions:
> 
> * geoip_table.ip_start_end <= ip_address <= geoip_table.ip_range_end

> * Regarding @LantaoJin's suggestions to use `Partitioner`, I am not to sure how to integrate this given the current design. Currently my guess is that each node of the partition would be a different sub range, and we would query sub range directly to reduce load on join? Is this the right interpretation or am I missing something?

The integer range from `ip_start_start` to `ip_start_end` between each CIDR notation should be ordering and no overlapping. Since Spark has a `RangePartitioning` default implementation, the Join operator with join condition `geoip_table.ip_start_start <= ip_address AND ip_address <= geoip_table.ip_range_end` would perform `RangeJoin` in optimal. But I realized there is **no** range join implementation in Apache Spark and EMR Spark currently. (Databricks Runtime has supported it https://docs.databricks.com/en/optimizations/range-join.html). So the join operator will perform `NestedLoopJoin` finally. So I think the performance won't be good but no necessary to consider how to leverage `RangePartitioning` in right now.

andy-k-improving commented on ppl-spark-integration/src/main/java/org/opensearch/sql/ast/AbstractNodeVisitor.java:
```diff
@@ -333,9 +338,14 @@ public T visitExistsSubquery(ExistsSubquery node, C context) {
   public T visitWindow(Window node, C context) {
     return visitChildren(node, context);
   }
+
```
?

normanj-bitquill replied:
I think that adding the blank line here is better formatting.

andy-k-improving commented on ppl-spark-integration/src/main/java/org/opensearch/sql/ast/tree/GeoIp.java:
```diff
@@ -0,0 +1,40 @@
+package org.opensearch.sql.ast.tree;
```
License header

andy-k-improving commented on ppl-spark-integration/src/main/java/org/opensearch/sql/ast/tree/GeoIp.java:
```diff
@@ -0,0 +1,40 @@
+import java.util.Arrays;
+import java.util.List;
+
+@ToString
```
I couldn't find the use age of ToString and Getter in this codebase.

normanj-bitquill commented on ppl-spark-integration/src/main/antlr4/OpenSearchPPLParser.g4:
```diff
@@ -392,6 +392,11 @@ sortbyClause
 
 evalClause
    : fieldExpression EQUAL expression
+   | geoipCommand
```
How about standardizing this a bit more. Such as:
```
evalClause
   : fieldExpression EQUAL evalClauseExpression
   ;
evalClauseExpression
   : expression
   | GEOIP LT_PRTHNS ipAddress = functionArg (COMMA properties = geoIpPropertyList)? RT_PRTHNS
   ;
```

Hopefully this would mean that you no longer need to change the `Eval` class, since it will always contain LET expressions.

kenrickyap replied:
I couldn't find a way to represent a join using a LET as LET is treated as an Expressions and can no longer effect the Logical Plan.

kenrickyap replied:
LET is refers to an expression tree node type that EVAL uses by default to represent child nodes. This node type is not able to rewrite the AstTree into a join as it is an expression and thus cannot influence the logical plan. As GeoipCommand must be a child of Eval and modify the AstTree into a join it does not use the Let node type but is rather an UnresolvedPlan which can rewrite the AstTree.

Have added comment to explain this

normanj-bitquill commented on ppl-spark-integration/src/main/java/org/opensearch/sql/ppl/utils/GeoipCatalystUtils.java:
```diff
@@ -0,0 +1,169 @@
+import static org.opensearch.sql.ppl.utils.DataTypeTransformer.seq;
+import static org.opensearch.sql.ppl.utils.JoinSpecTransformer.join;
+
+public interface GeoipCatalystUtils {
```
Can this get a name that doesn't include `Utils`, such as `GeoIpCatalystLogicalPlanGenerator`?

YANG-DB replied:
Or maybe `GeoIpCatalystLogicalPlanTranslator`

YANG-DB commented on ppl-spark-integration/src/main/java/org/opensearch/sql/expression/function/SerializableUdf.java:
```diff
@@ -8,10 +8,14 @@
+import java.net.InetAddress;
+import java.net.UnknownHostException;
 
 public interface SerializableUdf {
```
@kenrickyap can we rename the method in this class to be indicative of being related to the `geoiplocation` feature ?
maybe add an internal static class ?

kenrickyap replied:
done

YANG-DB commented on ppl-spark-integration/src/main/java/org/opensearch/sql/ppl/utils/GeoipCatalystUtils.java:
```diff
@@ -0,0 +1,169 @@
+        return applyProjection(parameters.getField(), parameters.getProperties(), context);
+    }
+
+    static LogicalPlan applyJoin(Expression ipAddress, CatalystPlanContext context) {
```
please add some javadoc explanations

kenrickyap replied:
added docs

YANG-DB commented on ppl-spark-integration/src/main/java/org/opensearch/sql/ppl/utils/GeoipCatalystUtils.java:
```diff
@@ -0,0 +1,169 @@
+    static private List<Expression> createGeoIpStructFields(List<String> attributeList) {
+        List<String> attributeListToUse;
+        if (attributeList == null || attributeList.isEmpty()) {
+            attributeListToUse = List.of(
```
can the strings here be extracted into static final const?

kenrickyap replied:
done

YANG-DB commented on ppl-spark-integration/src/main/java/org/opensearch/sql/ppl/parser/AstExpressionBuilder.java:
```diff
@@ -445,6 +444,36 @@ public UnresolvedExpression visitLambda(OpenSearchPPLParser.LambdaContext ctx) {
+            for (OpenSearchPPLParser.GeoIpPropertyContext property : ctx.geoIpProperty()) {
+                String propertyName;
+                if (property.COUNTRY_ISO_CODE() != null) {
+                    propertyName = "COUNTRY_ISO_CODE";
```
can we make this enum ?

kenrickyap replied:
done

kenrickyap commented on ppl-spark-integration/src/main/antlr4/OpenSearchPPLLexer.g4:
```diff
@@ -416,9 +416,6 @@ ISPRESENT:                          'ISPRESENT';
 CIDRMATCH:                          'CIDRMATCH';
 
-// Geo Loction 
-GEOIP:                              'GEOIP';
```
It was re-added in another location

kenrickyap commented on README.md:
````diff
@@ -75,7 +75,8 @@ bin/spark-shell --packages "org.opensearch:opensearch-spark-standalone_2.12:0.7.
 To build and run this PPL in Spark, you can run (requires Java 11):
 
 ```
-sbt clean sparkPPLCosmetic/publishM2
````
Added back

YANG-DB commented on ppl-spark-integration/src/main/java/org/opensearch/sql/ppl/CatalystQueryPlanVisitor.java:
```diff
@@ -539,19 +539,64 @@ public LogicalPlan visitRename(Rename node, CatalystPlanContext context) {
+            // build the plan with the projection step
+            return context.apply(p -> new org.apache.spark.sql.catalyst.plans.logical.Project(projectExpressions, p));
+        } else {
+            return context.apply(p -> p);
```
why no just use `context.getPlan()` ?

YANG-DB commented on docs/ppl-lang/planning/ppl-geoip.md:
```diff
@@ -0,0 +1,39 @@
+        - use the "abc" geoip datasource, retrieve only city, lat and lon
+
+
+2. **Proposed wiring with the geoip database**
```
please change the wording from `Proposed` to implementation.

YANG-DB reviewed (commented):
@kenrickyap 
please add the explanation of how the geoIP is resolved using the join and detail the future plan for additional data-sources to use 
IMO it should be added in the `ppl-geoip.md` doc

YANG-DB commented:
@kenrickyap please also add the  `Signed-off-by: Author Name <authoremail@example.com>` in every commit message.

YANG-DB commented:
@kenrickyap can u plz fix the DCO ?(developer certification)
thanks

andy-k-improving commented:
@kenrickyap If this is ready, can. you remove the Draft tag also?

kenrickyap commented:
@LantaoJin would you be able to provide a review?

andy-k-improving commented on ppl-spark-integration/src/main/antlr4/OpenSearchPPLLexer.g4:
```diff
@@ -428,6 +425,18 @@ TYPEOF:                             'TYPEOF';
+CONTINENT_NAME:                     'CONTINENT_NAME';
+REGION_ISO_CODE:                    'REGION_ISO_CODE';
+REGION_NAME:                        'REGION_NAME';
+CITY_NAME:                          'CITY_NAME';
```
`time_zone` seems missing from the list, as per `ppl-ip.md` described.

kenrickyap replied:
time_zone is already set in other part of lexer, would it be preferred to create a new Alias for geoip specifically?

andy-k-improving replied:
If that is the case, it's ok.

andy-k-improving commented:
@kenrickyap I only conducted the review for half of the changes, will continue tomorrow, also checkstyle is failing on CI, would you mind to fix?
Thanks,

YANG-DB commented:
@kenrickyap can u see why the build is failing ?

kenrickyap commented on docs/ppl-lang/planning/ppl-geoip-command.md:
````diff
@@ -0,0 +1,68 @@
+- an example of using `geoip` is equivalent to running the following SQL query:
+
+   ```SQL
+        SELECT source.*, struct(geoip.country_name, geoip.city_name) AS a
````
yes, if one property is provided a string will be returned, else a struct will be returned

kenrickyap replied:
added additional plan

YANG-DB commented on ppl-spark-integration/src/test/scala/org/opensearch/flint/spark/ppl/PPLLogicalPlanGeoipFunctionTranslatorTestSuite.scala:
```diff
@@ -28,20 +28,18 @@ class PPLLogicalPlanGeoipFunctionTranslatorTestSuite
+      ipAddress: UnresolvedAttribute,
+      left: LogicalPlan,
+      right: LogicalPlan): LogicalPlan = {
     val is_ipv4 = ScalaUDF(
```
@kenrickyap please change the usage of `ScalaUDF` to call the [`ScalaUDF.visit(...)`](https://github.com/opensearch-project/opensearch-spark/blob/957de4e27d3a243ccb9beccd2ee72ab231e9dcf4/ppl-spark-integration/src/main/java/org/opensearch/sql/expression/function/SerializableUdf.java#L151)

kenrickyap replied:
addressed

YANG-DB commented on ppl-spark-integration/src/test/scala/org/opensearch/flint/spark/ppl/PPLLogicalPlanGeoipFunctionTranslatorTestSuite.scala:
@kenrickyap please also add tests that have `fields ...` in the query and test behaviour of the following cases:
 - fields match exactly `geoip` requested fields
 - fields missing some `geoip` requested fields
 - fields contains fields from table that collide with `geoip` requested fields
 -

kenrickyap replied:
added above test cases to PPLLogicalPlanGeoipFunctionTranslatorTestSuite except for "fields contains fields from table that collide with geoip requested fields" as users table has not predefined schema. instead have created integ-test in FlintSparkPPLGeoipITSuite to test this case.

kenrickyap commented on integ-test/src/integration/scala/org/opensearch/flint/spark/ppl/FlintSparkPPLGeoipITSuite.scala:
```diff
@@ -0,0 +1,93 @@
+    }
+  }
+
+  test("test geoip with no parameters") {
```
would this be redundant as logical plan is already tested in integration tests within ppl project?

andy-k-improving replied:
Nope, it's similar but not identical, unit-test is purposed to assert the logical plan itself, but not the correctness of the result, as there is no runtime-spark being involved.

On the other side, integ-test is purposed, to first make sure the expected logical plan is being produced AND the ideal results being generated by the given logical-plan.

By considering this, any future code changes result in change of the logical plan should fail BOTH Unit and Integration test, but not either.

kenrickyap replied:
ahh I see thanks for the explanation :) have added logical plan check in integ-tests

YANG-DB replied:
@andy-k-improving great explanation !!

kenrickyap commented on docs/ppl-lang/planning/ppl-geoip-command.md:
```diff
@@ -0,0 +1,68 @@
+geoip function to add information about the geographical location of an IPv4 or IPv6 address
+
+**Implementation syntax**
+- `... | eval geoinfo = geoip([datasource,] ipAddress *[,properties])`
```
this is needed for future implementations of geoip, as this is part of the planning docs chose not to exclude. Would it be preferred to remove this reference?

andy-k-improving replied:
I reckon this is debatable, @YANG-DB would appreciate your thought on this.
Do we document what has been done (The fact), or what we think it should be (The ideal slide).
Personally I prefer to only document what got implemented thou (The fact), if indeed there will be another feature owner to implement source support in the future, I rather to have him || her to update this page or mark it as obsoleted, when the time comes.

kenrickyap replied:
my opinion is that we can remove datasource for now and if we need to alter the design later we can alter the current doc. To remain consistent it might be best to have datasource params supplied by config, however the main limitation of this implementation is that users can only leverage one datasource which is a use case that should be supported.

andy-k-improving commented on ppl-spark-integration/src/main/java/org/opensearch/sql/ppl/utils/GeoIpCatalystLogicalPlanTranslator.java:
```diff
@@ -0,0 +1,239 @@
+public interface GeoIpCatalystLogicalPlanTranslator {
+    String SPARK_CONF_KEY = "spark.geoip.tablename";
+    String DEFAULT_GEOIP_TABLE_NAME = "geoip";
+    String SOURCE_TABLE_ALIAS = "t1";
```
The name seems to be too generic, will this cause error when user already has a table call `t1` in the PPL command?
You can probably add a prefix `__GEO_IP__` or something similar to avoid the conflict.

kenrickyap replied:
it won't as these aliases are added on top of pre-existing table names (e.g. t1.t1 or t1.t2 if table name is t1 or t2), and aliases are dropped after join happens so any future logical plan that uses this plan will not be confused. There is a test case of this case "test geoip function - ipAddress col exist in geoip table"

andy-k-improving commented on ppl-spark-integration/src/main/java/org/opensearch/sql/ppl/utils/GeoIpCatalystLogicalPlanTranslator.java:
```diff
@@ -0,0 +1,239 @@
+    String DEFAULT_GEOIP_TABLE_NAME = "geoip";
+    String SOURCE_TABLE_ALIAS = "t1";
+    String GEOIP_TABLE_ALIAS = "t2";
+    List<String> GEOIP_TABLE_COLUMNS = List.of(
```
Just a suggestion, can we construct this list from the enum on Line:229 instead.
Or else in the future when we update one, we may missed the other. 
```
enum GeoIpProperty {
        COUNTRY_ISO_CODE,
        COUNTRY_NAME,
        CONTINENT_NAME,
        REGION_ISO_CODE,
        REGION_NAME,
        CITY_NAME,
        TIME_ZONE,
        LOCATION
    }
```

andy-k-improving commented on ppl-spark-integration/src/main/java/org/opensearch/sql/ppl/utils/GeoIpCatalystLogicalPlanTranslator.java:
```diff
@@ -0,0 +1,239 @@
+                    new And(
+                            new GreaterThanOrEqual(
+                                    getIpInt(ipAddress),
+                                    UnresolvedAttribute$.MODULE$.apply(seq(GEOIP_TABLE_ALIAS,"ip_range_start"))
```
`ip_range_start` should be a constant, similar to `String DEFAULT_GEOIP_TABLE_NAME = "geoip"`

kenrickyap replied:
addressed

andy-k-improving commented on ppl-spark-integration/src/main/java/org/opensearch/sql/ppl/utils/GeoIpCatalystLogicalPlanTranslator.java:
```diff
@@ -0,0 +1,239 @@
+                            ),
+                            new LessThan(
+                                    getIpInt(ipAddress),
+                                    UnresolvedAttribute$.MODULE$.apply(seq(GEOIP_TABLE_ALIAS,"ip_range_end"))
```
same

kenrickyap replied:
addressed

andy-k-improving commented on ppl-spark-integration/src/main/java/org/opensearch/sql/ppl/utils/GeoIpCatalystLogicalPlanTranslator.java:
```diff
@@ -0,0 +1,239 @@
+                    ),
+                    new EqualTo(
+                            getIsIpv4(ipAddress),
+                            UnresolvedAttribute$.MODULE$.apply(seq(GEOIP_TABLE_ALIAS,"ipv4"))
```
same.

kenrickyap replied:
addressed

andy-k-improving commented on ppl-spark-integration/src/main/java/org/opensearch/sql/ppl/utils/GeoIpCatalystLogicalPlanTranslator.java:
```diff
@@ -0,0 +1,239 @@
+
+        List<Expression> dropList = createGeoIpStructFields(new ArrayList<>());
+        dropList.addAll(List.of(
+                UnresolvedAttribute$.MODULE$.apply(seq(GEOIP_TABLE_ALIAS,"cidr")),
```
Same, to abstract the constant, and re-use for `ip_range_x`

kenrickyap replied:
addressed

andy-k-improving commented on ppl-spark-integration/src/main/java/org/opensearch/sql/ppl/utils/GeoIpCatalystLogicalPlanTranslator.java:
```diff
@@ -0,0 +1,239 @@
+                seq(ipAddress),
+                seq(),
+                Option.empty(),
+                Option.apply("ip_to_int"),
```
Same for the function name

kenrickyap replied:
I will address this commit in follow-up PR as these function names are linked to definition in SerializedUdf class. Plan is to convert SerializedUdf.visit to use enum instead of strings for function names. However as this would impact code not relevant to geoip will do this in seperate PR.

andy-k-improving commented on docs/ppl-lang/planning/ppl-geoip-command.md:
```diff
@@ -0,0 +1,68 @@
+- `... | eval geoinfo = geoip([datasource,] ipAddress *[,properties])`
+- generic syntax     
+- `... | eval geoinfo = geoip(ipAddress)`
+- use the default geoip datasource
```
If we decide to keep the multiple DS support on the doc, please ignore this.

kenrickyap replied:
removed datasource for now

kenrickyap commented on integ-test/src/integration/scala/org/opensearch/flint/spark/ppl/FlintSparkPPLGeoipITSuite.scala:
```diff
@@ -0,0 +1,93 @@
+
+    // Compare the results
+    implicit val rowOrdering: Ordering[Row] = Ordering.by[Row, String](_.getAs[String](0))
+    assert(results.sorted.sameElements(expectedResults.sorted))
```
added integ-test to check that ParseException is thrown if invalid property is provided

kenrickyap commented on docs/ppl-lang/planning/ppl-geoip-command.md:
```diff
@@ -0,0 +1,68 @@
+- use the "abc" geoip datasource
+- `... | eval geoinfo = geoip(ipAddress, city, location)`
+-  use the default geoip datasource, retrieve only city, and location
+- `... | eval geoinfo = geoip("abc", ipAddress, city, location")`
```
removed datasource for now

YANG-DB commented on docs/ppl-lang/planning/ppl-geoip-command.md:
```diff
@@ -0,0 +1,74 @@
+    - OpenSearch geospatial client - once geospatial client is published we can leverage client to utilize OpenSearch geo2ip functionality.
+- Additional datasource connection params will be provided through spark config options.
+
+### New syntax definition in ANTLR
```
@kenrickyap we can remove this part since its more implementation details

kenrickyap replied:
removed

YANG-DB commented on ppl-spark-integration/src/main/java/org/opensearch/sql/ppl/CatalystQueryPlanVisitor.java:
```diff
@@ -562,19 +565,63 @@ public LogicalPlan visitRename(Rename node, CatalystPlanContext context) {
+        List<Node> expressions = node.getExpressionList();
+
+        // Geoip function modifies logical plan and is treated as QueryPlanVisitor instead of ExpressionVisitor
+        for (Node expr : expressions) {
```
@kenrickyap how about the next code:
```
for (Node expr : expressions) {
    switch (expr) {
        case Let let -> {
            Alias alias = new Alias(let.getVar().getField().toString(), let.getExpression());
            aliases.add(alias);
        }
        case UnresolvedPlan unresolvedPlan -> unresolvedPlan.accept(this, context);
        default -> throw new SyntaxCheckException("Unexpected node type when visiting EVAL");
    }
}
```

YANG-DB replied:
oh ok sorry

YANG-DB commented on ppl-spark-integration/src/main/java/org/opensearch/sql/ppl/CatalystQueryPlanVisitor.java:
```diff
@@ -562,19 +565,63 @@ public LogicalPlan visitRename(Rename node, CatalystPlanContext context) {
-            // Create an UnresolvedStar for all-fields projection
-            context.getNamedParseExpressions().push(UnresolvedStar$.MODULE$.apply(Option.<Seq<String>>empty()));
+
+        if (!aliases.isEmpty()) {
```
can this alias related code be moved up to the part which created the alias ?

kenrickyap replied:
this must happen after the for loop as it is the for loop that populates aliases by processing the child nodes of EVAL

YANG-DB commented on ppl-spark-integration/src/main/java/org/opensearch/sql/ppl/parser/AstExpressionBuilder.java:
```diff
@@ -450,6 +451,26 @@ public UnresolvedExpression visitLambda(OpenSearchPPLParser.LambdaContext ctx) {
+                String propertyName = property.getText().toUpperCase();
+
+                try {
+                    GeoIpCatalystLogicalPlanTranslator.GeoIpProperty.valueOf(propertyName);
```
can you extract this validation code into a single line ? 
create a static function and call its

kenrickyap replied:
addressed

YANG-DB reviewed (commented):
@kenrickyap I've added a few comments
plz review - thanks

kenrickyap commented:
> @kenrickyap I've added a few comments plz review - thanks

@YANG-DB have addressed new PR comments, not sure if I fully understand some of the comments, have added my understanding of unaddressed comments

YANG-DB reviewed (approved):
OK LGTM !

andy-k-improving commented:
Thx for the taking the time to address comments. :)
